secedgar/filings/_base.py

```diff
@@ -1,12 +1,72 @@
 from abc import ABC, abstractmethod
+from secedgar.filings.filing_extractor import FilingExtractor
 import string
+import os
+import sys
+import asyncio
+from secedgar.utils import make_path
+from secedgar.utils.exceptions import EDGARQueryError
+from secedgar.client.aiohttp_client import RateLimitedClientSession
+from aiohttp import ClientSession, TCPConnector
+import importlib.util
+import time
+# import tqdm if possible
+tqdm_spec = importlib.util.find_spec('tqdm')
+tqdm = importlib.util.module_from_spec(tqdm_spec)
+sys.modules['tqdm'] = tqdm
+tqdm_spec.loader.exec_module(tqdm)
 
 
 class AbstractFiling(ABC):
     """Abstract base class for all SEC EDGAR filings.
 
     .. versionadded:: 0.1.5
     """
+    @property
+    @abstractmethod
+    def rate_limit(self):
+        """Passed to child classes."""
+        pass
+
+    async def wait_for_download_async(self, inputs):
+        """Asynchronously download links into files using rate limit."""
+        time.sleep(1)
+
+        async def fetch_and_save(link, path, session):
+            async with await session.get(link) as response:
+                # print(response.headers['Content-Length'])
+                contents = await response.read()
+                if contents.startswith(b'<!DOCTYPE'):
+                    raise EDGARQueryError("You hit the rate limit")
```

Comment:
we are requesting a text file. If we are served with a html file, it is the html document saying that we hit the rate limit

Reply:
Ok, got it. Thanks!

Reply:
```suggestion
                    # Raise error if given html instead of expected txt file
                    raise EDGARQueryError("You hit the rate limit")
```